### PR TITLE
Bouncy castle (DEV)

### PR DESCRIPTION
**Pull request by jurajkusnier**
- introduce a few methods helping with signature checks of certificates

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/utils/SignatureUtils.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+fun validateSignature(key: PublicKey, data: ByteArray, signature: ByteArray, algorithm: String) {
+    val javaSignature = if (algorithm.endsWith("withECDSA")) {
+        val (r, s) = signature.splitHalves()
+        DERSequence(
+            arrayOf(
+                ASN1Integer(BigInteger(1, r)),
+                ASN1Integer(BigInteger(1, s)),
+            )
+        ).encoded
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/utils/SignatureUtils.kt:35]: the split logic

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/utils/SignatureUtils.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        initVerify(key)
+        update(data)
+    }
+    if (!verifier.verify(javaSignature)) {
+        throw SignatureException()
+    }
+}
+
+private fun ByteArray.splitHalves(): Pair<ByteArray, ByteArray> =
+    take(size / 2).toByteArray() to drop(size / 2).toByteArray()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/utils/SignatureUtils.kt:49]: and here

**Review by mtwalli — COMMENTED** [review]
I'm not sure if we need this code at all, I just used a few snippets as follows:

File: Corona-Warn-App/build.gradle [hunk 3]
```diff
@@ -455,6 +455,7 @@ dependencies {
     // SECURITY
     implementation "androidx.security:security-crypto:1.0.0-rc03"
     implementation 'org.conscrypt:conscrypt-android:2.5.1'
+    implementation 'org.bouncycastle:bcpkix-jdk15on:1.56'
```
Comment (d4rken) [inline Corona-Warn-App/build.gradle:458]: Why `1.56` specifically? It's from 2016:
https://mvnrepository.com/artifact/org.bouncycastle/bcpkix-jdk15on

**Review by d4rken — COMMENTED** [review]
Would it be useful to make these methods part of a class instead of global methods to facilitate easier mocking in tests?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/utils/PemUtils.kt [hunk 4]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/utils/PemUtils.kt:1]: Do we need to insert BouncyCastle as security provider?
```
Security.insertProviderAt(....)
```

**Comment by jurajkusnier** [comment]
Closed in favor of https://github.com/corona-warn-app/cwa-app-android/pull/3779

**Pull request closed without merging**